x-pack/plugins/observability/server/domain/services/date_range.ts

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { assertNever } from '@kbn/std';
+import moment from 'moment';
+
+import type { TimeWindow } from '../../types/models/time_window';
+import {
+  calendarAlignedTimeWindowSchema,
+  DurationUnit,
+  rollingTimeWindowSchema,
+} from '../../types/schema';
+
+export interface DateRange {
+  from: Date;
+  to: Date;
+}
+
+export const toDateRange = (timeWindow: TimeWindow, currentDate: Date = new Date()): DateRange => {
+  if (calendarAlignedTimeWindowSchema.is(timeWindow)) {
+    const unit = toMomentUnitOfTime(timeWindow.duration.unit);
+    const now = moment.utc(currentDate).startOf('minute');
+    const startTime = moment.utc(timeWindow.calendar.start_time);
+
+    const differenceInUnit = now.diff(startTime, unit);
+    if (differenceInUnit < 0) {
+      throw new Error('Cannot compute date range with future starting time');
+    }
+
+    const from = startTime.clone().add(differenceInUnit, unit);
+    const to = from.clone().add(timeWindow.duration.value, unit);
+
+    return { from: from.toDate(), to: to.toDate() };
+  }
+
+  if (rollingTimeWindowSchema.is(timeWindow)) {
+    const unit = toMomentUnitOfTime(timeWindow.duration.unit);
+    const now = moment(currentDate).startOf('minute');
+
+    return {
+      from: now.clone().subtract(timeWindow.duration.value, unit).toDate(),
+      to: now.toDate(),
+    };
+  }
+
+  assertNever(timeWindow);
```

Comment:
Out of curiosity: what will happen if for some reason code reaches here?

Reply:
This function throws an error. 
But this function is mostly here for getting a type error at compile time if we add a new timewindow type that is not handled here. If you have the code checked out locally, try removing one of the `if` block, and you'll see a TS error

<img width="868" alt="image" src="https://user-images.githubusercontent.com/1376800/196461144-05225fc3-0607-40a2-bd06-233f95fa0b47.png">

Reply:
Interesting!